readme: update for new algorithm, add db table, add primer (#40)
datafold/data-diff: README.md

# **data-diff**

**data-diff is currently under heavy development, if you run into issues,
please file an issue and we'll help you out ASAP!**

**data-diff** is a command-line tool and Python library to efficiently diff
rows across two different databases.

* 🪢 Verifies across [many different databases][1] (e.g. Postgres -> Snowflake)
* 🔍 Outputs diff of rows in detail
* 🚨 Simple CLI/API to create monitoring and alerts
* 🔥 Verify 25M+ rows in less than 10s
* ♾️  Works for tables with 10s of billions of rows

**data-diff** splits the table into smaller segments, then checksums each
segment in both databases. When the checksums for a segment aren't equal, it
will further divide that segment into yet smaller segments, cheksumming those

> until it gets to the offending row(s). See [Technical Explanation][2] for more

Reviewer: Re "offending", I would stick to something less picturesque, like differing, or changed.

details.

> This approach has similar performance to `count(*)`, but is able to output each

Reviewer: Only true when there are no changes. Also, maybe worth mentioning threading here?

differing row. By pushing the compute into the databases, it's _much_ faster

> than querying for and comparing each row.

Reviewer: maybe "every row" instead of "each row"

# **data-diff**

**data-diff is currently under heavy development, if you run into issues,
please file an issue and we'll help you out ASAP!**

**data-diff** is a command-line tool and Python library to efficiently validate
whether two different databases are in sync, e.g. Postgres and Redshift. The
common use-case is to verify replication between two datastores with e.g.
Fivetran, Debezium, .. and fix missing rows and/or alert on missing/incorrect
data so you know whether something is wrong before your customer.

> **data-diff** doesn't pull down and compare every row, because it is extremely

Reviewer 2: It kinda sounds like you are calling data-diff slow, lol.

```suggestion
**data-diff** doesn't pull down and compare every row, because that would be extremely
```

│       Postgres       │              │      Snowflake       │
├──────────────────────┤              ├──────────────────────┤
│      id=1..100k      │              │      id=1..100k      │
├──────────────────────┤              ├──────────────────────┤
│    id=100k..200k     │              │    id=100k..200k     │
├──────────────────────┤              ├──────────────────────┤
│    id=200k..300k     ├─────────────▶│    id=200k..300k     │
├──────────────────────┤              ├──────────────────────┤
│    id=300k..400k     │              │    id=300k..400k     │
├──────────────────────┤              ├──────────────────────┤
│         ...          │              │         ...          │
├──────────────────────┤              ├──────────────────────┤
│      900k..100k      │              │      900k..100k      │
└───────────────────▲──┘              └▲─────────────────────┘
                    ┃                  ┃
                    ┃                  ┃
                    ┃ checksum queries ┃
                    ┃                  ┃
                  ┌─┻──────────────────┻────┐
                  │        data-diff        │
                  └─────────────────────────┘
```

> Now **data-diff** will start running `--threads=1` queries in parallel that

Reviewer 2: Don't forget to add `--threads` to the options reference up above in the "how to use" section.

Author: Yep, we're just revising the wording for this in another PR — where we'll do this 👍🏻 

We want `--threads=1` to mean "1 simultaneous query per database", today it doesn't really mean that...

Options:

  - `--help` - Show help message and exit.
  - `-k` or `--key-column` - Name of the primary key column
  - `-t` or `--update-column` - Name of updated_at/last_updated column
  - `-c` or `--columns` - List of names of extra columns to compare
  - `-l` or `--limit` - Maximum number of differences to find (limits maximum bandwidth and runtime)
  - `-s` or `--stats` - Print stats instead of a detailed diff
  - `-d` or `--debug` - Print debug info
  - `-v` or `--verbose` - Print extra info
  - `-i` or `--interactive` - Confirm queries, implies `--debug`
  - `--max-age` - Considers only rows younger than specified. See `--min-age`.
                  Useful for specifying replication lag.
  - `--min-age` - Considers only rows older than specified.

>                   Example: `--min-age=5min` considers only rows from the last 5 minutes.

Reviewer: It's ignore, not considers.

It looks like it's really confusing, so maybe we should re-think the name or phrasing?

Author: Hmmmm yeah we definitely need to rethink it, but isn't this right?

Drew a diagram below... Here the `--min-age` is the `id >= now() -5min` and `--max-age` is `id <= now() - 2min`?

```
┌ ─ ─ ┬─────┬─────┬─────┬─────┬ ─ ─ ┬ ─ ─ ┬ ─ ─ ┐
 11:59│12:00│12:01│12:02│12:03│12:04 12:05 12:06
└ ─ ─ ┴──▲──┴─────┴─────┴──▲──┴ ─ ─ ┴ ─ ▲ ┴ ─ ─ ┘
         │                 │            │
         │                 │            │
         │                 │            │
  ┌──────┴───────┐    ┌────┴─────────┐  │ ┌─────────────┐
  │ --min-age=5m │    │ --max-age=2m │  └─┤Present time │
  └──────────────┘    └──────────────┘    └─────────────┘
```

Reviewer: No, min-age at this time means anything before 12.

And a min-age that's bigger than the max-age shouldn't be allowed.

When you go back in time the "time value" or timestamp get smaller, that's true, but the age actually gets bigger. (like someone born in the 80s is "bigger" than someone from the 90s, despite the number being smaller)

The "age" of 12:00 is 5 mins. The age of 12:01 is 4 mins, which is below 5, aka the minimum age, so it's out.

The "age" of 11:00 is an hour and 5 mins, which is a bigger age than the minimum age.

Does it make sense?

Author: Yeah, but I mean... I drew it, and I'm still confused 😓  

But I think we need to change it to `--min-updated-key` and `--max-updated-key` which would line up with the drawing above

Reviewer: But then it's not very clear that the value is relative to now. It sounds more like an absolute value..

Author: Yeah, but I think it's clearer then if you do `--min-updated -5min` or something.

Reviewer 2: Hi Y'all, end user here.  I think the semantics of “age” matter a lot.  The source of confusion is that “age” is implicitly negative, but we think about it as a positive number/absolute value, and so inverts the conventional meaning of “min” and “max” when thinking about timestamps.  Your readme documents `--min-age` being a lower bound on “age” and an upper bound on “timestamp” (matching my intuition), but the implementation does the reverse.

Regardless, docs should match readme, and ideally should match user intuition as well.
